### PR TITLE
Change printStats on Multi-layer caches.

### PR DESCRIPTION
**Pull request by hangshao0**
Change printStats on Multi-layer caches. 

1. Change "printStats" to show statistics as top-layer statistics
and all-layer statistics. Change "printStat=" to show cache entries from
all layers. 

2. Add new options "printTopLayerStats" and "printTopLayerStats=", which
show statistics and cache entries from the top layer only.

3. Change the cache statistics in javacore to show as top-layer
statistics and all-layer statistics. 

4. For single layer cache, the order of entries in the output of
"printStats" and "printDetails" has been changed.

issue #5480

Signed-off-by: Hang Shao <hangshao@ca.ibm.com>

### CODE REVIEW

**Comment by hangshao0** [comment]
Result on cache with layer 0 and layer 1:

```
java -Xshareclasses:name=Cache1,printStats

Current statistics for top layer of cache "Cache1":

Cache created with:
        -Xnolinenumbers                      = false
        BCI Enabled                          = true
        Restrict Classpaths                  = false
        Feature                              = cr


base address                         = 0x00007F70D4529000
end address                          = 0x00007F70D550D000
allocation pointer                   = 0x00007F70D4529000

cache layer                          = 1
cache size                           = 16776608
softmx bytes                         = 16776608
free bytes                           = 15322100
Reserved space for AOT bytes         = -1
Maximum space for AOT bytes          = -1
Reserved space for JIT data bytes    = -1
Maximum space for JIT data bytes     = -1
Class debug area size                = 1331200
Class debug area used bytes          = 0
Class debug area % used              = 0%

Cache is 8% full

Cache is accessible to current user = true
---------------------------------------------------------

Current statistics for all layers of cache "Cache1":

ROMClass bytes                       = 1459040
AOT bytes                            = 54512
JIT data bytes                       = 456
Zip cache bytes                      = 1134016
Startup hint bytes                   = 0

# ROMClasses                         = 503
# AOT Methods                        = 15
# Classpaths                         = 1
# URLs                               = 0
# Tokens                             = 0
# Zip caches                         = 21
# Startup hints                      = 0
# Stale classes                      = 0
% Stale classes                      = 0%
```

```
java -Xshareclasses:name=Cache1,printtopLayerStats

Current statistics for cache "Cache1":

Cache created with:
        -Xnolinenumbers                      = false
        BCI Enabled                          = true
        Restrict Classpaths                  = false
        Feature                              = cr


base address                         = 0x00007FAF56CF6000
end address                          = 0x00007FAF57CDA000
allocation pointer                   = 0x00007FAF56CF6000

cache size                           = 16776608
softmx bytes                         = 16776608
free bytes                           = 15322100
ROMClass bytes                       = 0
AOT bytes                            = 8744
Reserved space for AOT bytes         = -1
Maximum space for AOT bytes          = -1
JIT data bytes                       = 108
Reserved space for JIT data bytes    = -1
Maximum space for JIT data bytes     = -1
Zip cache bytes                      = 0
Startup hint bytes                   = 0
Data bytes                           = 114080
Metadata bytes                       = 376
Metadata % used                      = 0%
Class debug area size                = 1331200
Class debug area used bytes          = 0
Class debug area % used              = 0%
stale bytes                          = 0

# ROMClasses                         = 0
# AOT Methods                        = 2
# Classpaths                         = 0
# URLs                               = 0
# Tokens                             = 0
# Zip caches                         = 0
# Startup hints                      = 0
# Stale classes                      = 0
% Stale classes                      = 0%

Cache is 8% full

Cache is accessible to current user = true
```

**Comment by hangshao0** [comment]
Javacore:
```

NULL           ------------------------------------------------------------------------
0SECTION       SHARED CLASSES subcomponent dump routine
NULL           ========================================
NULL
1SCLTEXTCSTL   Cache Statistics for Top Layer
NULL
1SCLTEXTCRTW       Cache Created With
NULL               ------------------
NULL
2SCLTEXTXNL            -Xnolinenumbers       = false
2SCLTEXTBCI            BCI Enabled           = true
2SCLTEXTBCI            Restrict Classpaths   = false
NULL
1SCLTEXTCSUM       Cache Summary
NULL               ------------------
NULL
2SCLTEXTNLC            No line number content                    = false
2SCLTEXTLNC            Line number content                       = false
NULL
2SCLTEXTRCS            ROMClass start address                    = 0x00007F0EDB567000
2SCLTEXTRCE            ROMClass end address                      = 0x00007F0EDB567000
2SCLTEXTMSA            Metadata start address                    = 0x00007F0EDC40241C
2SCLTEXTCEA            Cache end address                         = 0x00007F0EDC54B000
2SCLTEXTRTF            Runtime flags                             = 0x80102001ECA602BB
2SCLTEXTCGN            Cache generation                          = 41
NULL
2SCLTEXTCLY            Cache layer                               = 1
2SCLTEXTCSZ            Cache size                                = 16776608
2SCLTEXTSMB            Softmx bytes                              = 16776608
2SCLTEXTFRB            Free bytes                                = 15315996
2SCLTEXTARB            Reserved space for AOT bytes              = -1
2SCLTEXTAMB            Maximum space for AOT bytes               = -1
2SCLTEXTJRB            Reserved space for JIT data bytes         = -1
2SCLTEXTJMB            Maximum space for JIT data bytes          = -1
2SCLTEXTRWB            ReadWrite bytes                           = 114080
2SCLTEXTDAS            Class debug area size                     = 1331200
2SCLTEXTDAU            Class debug area % used                   = 0%
2SCLTEXTDAN            Class LineNumberTable bytes               = 0
2SCLTEXTDAV            Class LocalVariableTable bytes            = 0
NULL
2SCLTEXTCPF            Cache is 8% full
NULL
1SCLTEXTCMST       Cache Memory Status
NULL               ------------------
1SCLTEXTCNTD           Cache Name                    Feature                  Memory type              Cache path
NULL
2SCLTEXTCMDT           Cache1                        CR                       Memory mapped file       /tmp/javasharedresources/C290M4F1A64P_Cache1_G41L01
NULL
1SCLTEXTCMST       Cache Lock Status
NULL               ------------------
1SCLTEXTCNTD           Lock Name                     Lock type                TID owning lock
NULL
2SCLTEXTCWRL           Cache write lock              File lock                Unowned
2SCLTEXTCRWL           Cache read/write lock         File lock                Unowned
NULL
1SCLTEXTCSAL   Cache Statistics for All Layers
NULL
2SCLTEXTRCB            ROMClass bytes                            = 1459040
2SCLTEXTAOB            AOT code bytes                            = 57624
2SCLTEXTADB            AOT data bytes                            = 272
2SCLTEXTAHB            AOT class hierarchy bytes                 = 1840
2SCLTEXTATB            AOT thunk bytes                           = 632
2SCLTEXTJHB            JIT hint bytes                            = 484
2SCLTEXTJPB            JIT profile bytes                         = 0
2SCLTEXTNOB            Java Object bytes                         = 0
2SCLTEXTZCB            Zip cache bytes                           = 1134016
2SCLTEXTSHB            Startup hint bytes                        = 0
2SCLTEXTJCB            JCL data bytes                            = 0
2SCLTEXTBDA            Byte data bytes                           = 0
NULL
2SCLTEXTNRC            Number ROMClasses                         = 503
2SCLTEXTNAM            Number AOT Methods                        = 16
2SCLTEXTNAD            Number AOT Data Entries                   = 1
2SCLTEXTNAH            Number AOT Class Hierarchy                = 28
2SCLTEXTNAT            Number AOT Thunks                         = 11
2SCLTEXTNJH            Number JIT Hints                          = 15
2SCLTEXTNJP            Number JIT Profiles                       = 0
2SCLTEXTNCP            Number Classpaths                         = 1
2SCLTEXTNUR            Number URLs                               = 0
2SCLTEXTNTK            Number Tokens                             = 0
2SCLTEXTNOJ            Number Java Objects                       = 0
2SCLTEXTNZC            Number Zip Caches                         = 21
2SCLTEXTNSH            Number Startup Hint Entries               = 0
2SCLTEXTNJC            Number JCL Entries                        = 0
2SCLTEXTNST            Number Stale classes                      = 0
2SCLTEXTPST            Percent Stale classes                     = 0%
```

**Comment by pshipton** [comment]
The following are missing from layered cache stats.
```
Data bytes                           =
Metadata bytes                       = 
Metadata % used                      = 

stale bytes                          = 
```

And from the javacore.
```
2SCLTEXTMDA        Metadata bytes
```

**Comment by hangshao0** [comment]
`descriptor->otherBytes` was reported as metadata bytes. This is how it was calculated:

https://github.com/eclipse/openj9/blob/master/runtime/shared_common/CacheMap.cpp#L4426

In the multi-cache case, some of those fields are for top-layer cache and some are for all layers, so this calculation is not valid anymore. So I removed it from the multi-layer statistics. But user can still use printTopLayerStats to see the metadata bytes of the top layer if they are interested in this value.

File: runtime/rasdump/javadump.cpp [hunk 1]
```diff
@@ -2806,396 +2806,788 @@ JavaCoreDumpWriter::writeSharedClassSection(void)
...
 
+			writeSharedClassLockInfo(
+					"2SCLTEXTCWRL       Cache write lock              ", javacoreData.semid, javacoreData.writeLockTID
+			);
+		
+			writeSharedClassLockInfo(
+					"2SCLTEXTCRWL       Cache read/write lock         ", javacoreData.semid, javacoreData.readWriteLockTID
+			);
+		} else {
 			_OutputStream.writeCharacters(
```
Comment (pshipton) [inline runtime/rasdump/javadump.cpp]: How about putting some of this into helper functions so it's not all duplicated between the two different types of output. Similarly for SH_CacheMap::printCacheStats().

Comment (hangshao0) [reply runtime/rasdump/javadump.cpp]: Added 3 helper functions

**Comment by hangshao0** [comment]
Output of `printStats `and `printTopLayerStats `on a 2-layer shared cache:

```
java -Xshareclasses:name=Cache1,printStats

Current statistics for top layer of cache "Cache1":

Cache created with:
        -Xnolinenumbers                      = false
        BCI Enabled                          = true
        Restrict Classpaths                  = false
        Feature                              = cr

Cache contains only classes with line numbers

base address                         = 0x00007F394E20B000
end address                          = 0x00007F394F1EF000
allocation pointer                   = 0x00007F394E221EA0

cache layer                          = 1
cache size                           = 16776608
softmx bytes                         = 16776608
free bytes                           = 15228964
Reserved space for AOT bytes         = -1
Maximum space for AOT bytes          = -1
Reserved space for JIT data bytes    = -1
Maximum space for JIT data bytes     = -1
Class debug area size                = 1331200
Class debug area used bytes          = 7728
Class debug area % used              = 0%

Cache is 9% full

Cache is accessible to current user = true
---------------------------------------------------------

Current statistics for all layers of cache "Cache1":

ROMClass bytes                       = 1478240
AOT bytes                            = 38544
JIT data bytes                       = 416
Zip cache bytes                      = 1134016
Startup hint bytes                   = 0

# ROMClasses                         = 510
# AOT Methods                        = 15
# Classpaths                         = 1
# URLs                               = 0
# Tokens                             = 0
# Zip caches                         = 21
# Startup hints                      = 0
# Stale classes                      = 0
% Stale classes                      = 0%

java -Xshareclasses:name=Cache1,printTopLayerStats

Current statistics for cache "Cache1":

Cache created with:
        -Xnolinenumbers                      = false
        BCI Enabled                          = true
        Restrict Classpaths                  = false
        Feature                              = cr

Cache contains only classes with line numbers

base address                         = 0x00007F2812EEA000
end address                          = 0x00007F2813ECE000
allocation pointer                   = 0x00007F2812F00EA0

cache size                           = 16776608
softmx bytes                         = 16776608
free bytes                           = 15228964
Reserved space for AOT bytes         = -1
Maximum space for AOT bytes          = -1
Reserved space for JIT data bytes    = -1
Maximum space for JIT data bytes     = -1
Data bytes                           = 114080
Metadata bytes                       = 1664
Metadata % used                      = 0%
Class debug area size                = 1331200
Class debug area used bytes          = 7728
Class debug area % used              = 0%
ROMClass bytes                       = 93856
AOT bytes                            = 6768
JIT data bytes                       = 76
Zip cache bytes                      = 0
Startup hint bytes                   = 0
stale bytes                          = 0

# ROMClasses                         = 33
# AOT Methods                        = 2
# Classpaths                         = 0
# URLs                               = 0
# Tokens                             = 0
# Zip caches                         = 0
# Startup hints                      = 0
# Stale classes                      = 0
% Stale classes                      = 0%


Cache is 9% full

Cache is accessible to current user = true

```

**Comment by pshipton** [comment]
What about `Data bytes` and `stale bytes`?

**Comment by hangshao0** [comment]
`Data bytes` is calculated as: 
`indexedDataBytes + readWriteBytes + jclDataBytes + objectBytes`

The readWriteBytes is for top layer only and the rest are for all layers so I did not report `Data bytes` in multi-layer stats. On my second thought, as the readWriteBytes in lower layer won't be used, so we can report `Data bytes` in the multi-layer stats (change updated)

`stale bytes` shows up in `printStats=` and `printTopLayerStats=`. If you are print stats for all layers, it shows stale bytes of all layers. If you are print stats for top layer only, it is stale bytes for top layer.

**Comment by hangshao0** [comment]
To be consistent with the output of `printStats`, I updated the change to remove `stale bytes` from `printTopLayerStats`. It shows up in `printStats=` and `printTopLayerStats=`

File: runtime/shared_common/CacheMap.cpp [hunk 2]
```diff
@@ -4303,18 +4300,44 @@ SH_CacheMap::releasePrivateSharedData(J9VMThread* currentThread, const J9SharedD
 }
 
 /* THREADING: Can be called at any time by any thread. Should not try to get any locks as it may be being
- * called as a result of a deadlock. The only locks obtained are by the managers when querying their hashtables. */
+ * called as a result of a deadlock. The only locks obtained are by the managers when querying their hashtables.
+ * 
+ * Get the J9SharedClassJavacoreDataDescriptor from the cache of all layers.
```
Comment (pshipton) [inline runtime/shared_common/CacheMap.cpp]: This needs to be worded better. Like, Get the J9SharedClassJavacoreDataDescriptor for all layers of the cache.

File: runtime/shared_common/CacheMap.cpp [hunk 3]
```diff
@@ -4303,18 +4300,44 @@ SH_CacheMap::releasePrivateSharedData(J9VMThread* currentThread, const J9SharedD
...
+}
+
+/* THREADING: Can be called at any time by any thread. Should not try to get any locks as it may be being
+ * called as a result of a deadlock. The only locks obtained are by the managers when querying their hashtables. 
+ * 
+ * Get the J9SharedClassJavacoreDataDescriptor for the current shared cache.
+ * 
+ * @param[in] vm  The J9JavaVM
+ * @param[out] descriptor The J9SharedClassJavacoreDataDescriptor
+ * @param[in] topLayerOnly  Whether J9SharedClassJavacoreDataDescriptor the from the cache of top layer only or all layers
```
Comment (pshipton) [inline runtime/shared_common/CacheMap.cpp]: from the top layer cache only or all layers

File: runtime/shared_common/shrinit.cpp [hunk 4]
```diff
@@ -3276,8 +3303,11 @@ j9shr_init(J9JavaVM *vm, UDATA loadFlags, UDATA* nonfatal)
 	if (parseResult==RESULT_DO_PRINTSTATS ||
 		parseResult==RESULT_DO_PRINTALLSTATS ||
```
Comment (pshipton) [inline runtime/shared_common/shrinit.cpp]: might as well fix the formatting for all of the conditions

File: runtime/shared_common/shrinit.cpp [hunk 5]
```diff
@@ -3851,7 +3881,10 @@ j9shr_print_stats(J9JavaVM *vm, UDATA parseResult, U_64 runtimeFlags, UDATA prin
 			RESULT_DO_PRINTALLSTATS_EQUALS == parseResult||
```
Comment (pshipton) [inline runtime/shared_common/shrinit.cpp]: might as well fix the formatting for all of the conditions

**Comment by pshipton** [comment]
Pls provide the stats for regular cache, not multi-layer, before and after this change on the same cache file, to confirm the only changes are the order of the lines.

**Comment by hangshao0** [comment]
Before this change: 

```
Current statistics for cache "Cache1":

Cache created with:
        -Xnolinenumbers                      = false
        BCI Enabled                          = true
        Restrict Classpaths                  = false
        Feature                              = cr

Cache contains only classes with line numbers

base address                         = 0x00007F771784B000
end address                          = 0x00007F771882F000
allocation pointer                   = 0x00007F771798F2B0

cache size                           = 16776608
softmx bytes                         = 16776608
free bytes                           = 12780740
ROMClass bytes                       = 1327792
AOT bytes                            = 58408
Reserved space for AOT bytes         = -1
Maximum space for AOT bytes          = -1
JIT data bytes                       = 520
Reserved space for JIT data bytes    = -1
Maximum space for JIT data bytes     = -1
Zip cache bytes                      = 1133640
Startup hint bytes                   = 0
Data bytes                           = 114080
Metadata bytes                       = 30228
Metadata % used                      = 0%
Class debug area size                = 1331200
Class debug area used bytes          = 189326
Class debug area % used              = 14%

# ROMClasses                         = 452
# AOT Methods                        = 16
# Classpaths                         = 1
# URLs                               = 0
# Tokens                             = 0
# Zip caches                         = 21
# Startup hints                      = 0
# Stale classes                      = 0
% Stale classes                      = 0%

Cache is 23% full

Cache is accessible to current user = true
```

After this change;
```
Current statistics for cache "Cache1":

Cache created with:
        -Xnolinenumbers                      = false
        BCI Enabled                          = true
        Restrict Classpaths                  = false
        Feature                              = cr

Cache contains only classes with line numbers

base address                         = 0x00007F53169FE000
end address                          = 0x00007F53179E2000
allocation pointer                   = 0x00007F5316B422B0

cache size                           = 16776608
softmx bytes                         = 16776608
free bytes                           = 12780740
Reserved space for AOT bytes         = -1
Maximum space for AOT bytes          = -1
Reserved space for JIT data bytes    = -1
Maximum space for JIT data bytes     = -1
Metadata bytes                       = 30228
Metadata % used                      = 0%
Class debug area size                = 1331200
Class debug area used bytes          = 189326
Class debug area % used              = 14%
ROMClass bytes                       = 1327792
AOT bytes                            = 58408
JIT data bytes                       = 520
Zip cache bytes                      = 1133640
Startup hint bytes                   = 0
Data bytes                           = 114080

# ROMClasses                         = 452
# AOT Methods                        = 16
# Classpaths                         = 1
# URLs                               = 0
# Tokens                             = 0
# Zip caches                         = 21
# Startup hints                      = 0
# Stale classes                      = 0
% Stale classes                      = 0%


Cache is 23% full

Cache is accessible to current user = true
```

**Comment by pshipton** [comment]
I think the new output should have a blank line before `ROMClass bytes`. Makes it a bit easier to find this section now that it's not near the top.

File: runtime/shared_common/CacheMap.cpp (hunk 2 above)
Comment (hangshao0) [reply runtime/shared_common/CacheMap.cpp]: Fixed

File: runtime/shared_common/shrinit.cpp (hunk 4 above)
Comment (hangshao0) [reply runtime/shared_common/shrinit.cpp]: Fixed

File: runtime/shared_common/shrinit.cpp (hunk 5 above)
Comment (hangshao0) [reply runtime/shared_common/shrinit.cpp]: Fixed

File: runtime/shared_common/CacheMap.cpp (hunk 3 above)
Comment (hangshao0) [reply runtime/shared_common/CacheMap.cpp]: Fixed

**Comment by hangshao0** [comment]
> I think the new output should have a blank line before ROMClass bytes

Blank line added.

File: runtime/shared_common/shrinit.cpp [hunk 6]
```diff
@@ -3847,11 +3877,14 @@ j9shr_print_stats(J9JavaVM *vm, UDATA parseResult, U_64 runtimeFlags, UDATA prin
 	UDATA showFlags = 0;
 	J9VMThread* currentThread = vm->internalVMFunctions->currentVMThread(vm);
 
-	if (RESULT_DO_PRINTALLSTATS == parseResult ||
-			RESULT_DO_PRINTALLSTATS_EQUALS == parseResult||
-			RESULT_DO_PRINTORPHANSTATS == parseResult ||
-			RESULT_DO_PRINTSTATS_EQUALS == parseResult ||
-			RESULT_DO_PRINTSTATS == parseResult) {
+	if ((RESULT_DO_PRINTALLSTATS == parseResult) ||
+		(RESULT_DO_PRINTALLSTATS_EQUALS == parseResult)||
```
Comment (pshipton) [inline runtime/shared_common/shrinit.cpp]: missing space before `||`

File: runtime/shared_common/CompositeCache.cpp [hunk 7]
```diff
@@ -7073,3 +7073,12 @@ SH_CompositeCacheImpl::getCacheName(void) const
 {
 	return _cacheName;
 }
+
+/**
+ * Return the cache layer number
```
Comment (pshipton) [inline runtime/shared_common/CompositeCache.cpp]: Should end with a period.

File: runtime/shared_common/CacheMap.cpp [hunk 8]
```diff
@@ -5062,10 +5093,222 @@ SH_CacheMap::printAllCacheStats(J9VMThread* currentThread, UDATA showFlags, SH_C
 		}
 	} while (it); 
 
-	_ccHead->exitWriteMutex(currentThread, fnName);
+	cache->exitWriteMutex(currentThread, fnName);
 	return 0;
 }
 
+/*
+ * Helper funtion to print the statistics of the top layer cache
```
Comment (pshipton) [inline runtime/shared_common/CacheMap.cpp]: Should end with a period.

File: runtime/shared_common/CacheMap.cpp [hunk 9]
```diff
@@ -5062,10 +5093,222 @@ SH_CacheMap::printAllCacheStats(J9VMThread* currentThread, UDATA showFlags, SH_C
...
+	if (J9_ARE_ALL_BITS_SET(runtimeFlags, J9SHR_RUNTIMEFLAG_ENABLE_DETAILED_STATS)) {
+		CACHEMAP_FMTPRINT1(J9NLS_DO_NOT_PRINT_MESSAGE_TAG, J9NLS_SHRC_CM_PRINTSTATS_DEBUGAREA_LINENUMBERTABLE_BYTES, javacoreData->debugAreaLineNumberTableBytes);
+		CACHEMAP_FMTPRINT1(J9NLS_DO_NOT_PRINT_MESSAGE_TAG, J9NLS_SHRC_CM_PRINTSTATS_DEBUGAREA_LOCALVARIABLETABLE_BYTES_V2, javacoreData->debugAreaLocalVariableTableBytes);
+	} else {
+		CACHEMAP_FMTPRINT1(J9NLS_DO_NOT_PRINT_MESSAGE_TAG, J9NLS_SHRC_CM_PRINTSTATS_DEBUGAREA_USED_BYTES, javacoreData->debugAreaLineNumberTableBytes + javacoreData->debugAreaLocalVariableTableBytes);
+	}
+	CACHEMAP_FMTPRINT1(J9NLS_DO_NOT_PRINT_MESSAGE_TAG, J9NLS_SHRC_CM_PRINTSTATS_DEBUGAREA_USED, javacoreData->debugAreaUsed);
+}
+/*
+ * Helper funtion to print the statistics summary of the top layer cache
```
Comment (pshipton) [inline runtime/shared_common/CacheMap.cpp]: Should end with a period.

**Comment by pshipton** [comment]
jenkins test sanity,extended win jdk8

**Comment by pshipton** [comment]
https://ci.eclipse.org/openj9/job/PullRequest-OpenJ9/1385/

**Comment by pshipton** [comment]
PR testing successfully completed.

**Comment by pshipton** [comment]
@hangshao0 please cherry pick and create a PR for the v0.17.0-release branch.

**Pull request merged**